### DISCUSSION TITLE
How does it actually create binaries?

### DISCUSSION THREAD

**Discussion by Blank-c** [pyinstaller/pyinstaller#6995] [Help] [answered]
How does pyinstaller create an exe out of python files and modules? Does it have a c compiler comewhere? Please explain

**Comment by rokm** [comment, marked as answer]
> Does it have a c compiler comewhere?

Nope.

We have pre-built bootloader executable (for each OS and arch; or built locally if prebuilt variant is unavailable) that loads the collected python shared library and uses it to set up embedded python interpreter. Then, it loads and runs the collected byte-compiled python code from the embedded archive (which is appended to the executable). In onefile mode, the external data files and shared libs are also collected in executable-embedded archive; in that case, the parent bootloader process unpacks everything into temporary directory, then launches a child process (using the same executable) that sets up python interpreter and runs the python code.
